itests/hive-minikdc/src/test/java/org/apache/hive/minikdc/TestJdbWithCustomAuthWithKerberos.java

```diff
@@ -0,0 +1,193 @@
+  private static Integer KERBEROS_CUSTOM_PORT;
+  private static final String HS2_BINARY_MODE = "binary";
+  private static final String KEY_STORE_NAME = "keystore.jks";
+  private static final String TRUST_STORE_NAME = "truststore.jks";
+  private static final String KEY_STORE_PASSWORD = "HiveJdbc";
+  private static final String JAVA_TRUST_STORE_PROP = "javax.net.ssl.trustStore";
+  private static final String JAVA_TRUST_STORE_PASS_PROP = "javax.net.ssl.trustStorePassword";
+
+  private MiniHiveKdc miniHiveKdc;
+  private MiniHS2 miniHS2 = null;
+  private static HiveConf hiveConf = new HiveConf();
+  private Connection hs2Conn = null;
+  private Map<String, String> confOverlay;
+  private String dataFileDir = hiveConf.get("test.data.files");
+  private final String SSL_CONN_PARAMS = ";ssl=true;sslTrustStore=" + URLEncoder.encode(dataFileDir + File.separator +
+      TRUST_STORE_NAME) + ";trustStorePassword=" + KEY_STORE_PASSWORD;
+
+  @BeforeClass
+  public static void beforeTest() throws Exception {
+    Class.forName(MiniHS2.getJdbcDriverName());
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    DriverManager.setLoginTimeout(0);
+    KERBEROS_CUSTOM_PORT = MetaStoreUtils.findFreePort();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    if (hs2Conn != null) {
+      try {
+        hs2Conn.close();
+      } catch (Exception e) {
+        // Ignore shutdown errors since there are negative tests
+      }
+    }
+    if (miniHS2 != null && miniHS2.isStarted()) {
+      miniHS2.stop();
+    }
+    System.clearProperty(JAVA_TRUST_STORE_PROP);
+    System.clearProperty(JAVA_TRUST_STORE_PASS_PROP);
+  }
+
+  @Test
+  public void testKerberosAuthentication() throws Exception {
+    setCustomAuthWithKrbOverlay();
+    startMiniHS2();
+
+    // JDBC connection to HiveServer2 with Kerberos
+    hs2Conn = DriverManager.getConnection(miniHS2.getJdbcURL());
+    hs2Conn.close();
+  }
+
+  @Test
+  public void testCustomAuthenticationOverPlainWithKerberos() throws Exception {
+    setCustomAuthWithKrbOverlay();
+    startMiniHS2();
+
+    // JDBC connection with ID/PASSWD without SSL with Kerberos
+    String url = "jdbc:hive2://" + miniHS2.getHost() + ":" + KERBEROS_CUSTOM_PORT + "/default";
+
+    // wrong ID/PASSWD
+    try {
+      hs2Conn = DriverManager.getConnection(url, "wronguser", "pwd");
+    } catch(SQLException e) {
+      assertNotNull(e.getMessage());
+      assertTrue(e.getMessage(), e.getMessage().contains("Peer indicated failure: Error validating the login"));
+    }
+
+    // success ID/PASSWD
+    hs2Conn = DriverManager.getConnection(url, "hiveuser", "hive");
+    hs2Conn.close();
+
+    System.out.println(">>> PASSED testCustomAuthenticationOverPlain");
+    hs2Conn.close();
+  }
+
+  @Test
+  public void testCustomAuthenticationOverSSLWithKerberos() throws Exception {
+    setCustomAuthWithKrbOverlay();
+    setCustomAuthOverSslConfWithKrbOverlay(confOverlay);
+    startMiniHS2();
+
```

Comment:
from 118 to 136 this looks very similar to test code above (93-110).
make a shared function call?